### PR TITLE
Summarise visits by minute; buffer in memory and in CSV before DB

### PR DESCRIPTION
**Pull request by vladsavelyev**
- [x] Instead of storing individual visits, summarize them by minute by attributes
- [x] Rename `Visit` to `Visits` and use fields `start`, `end`, `count` instead of `timestamp`
- [x] Instead of directly pushing visits to DB, buffer them in memory, and every 10 sec persist in a CSV file, and daily summarize and persist in the DB
- [x] Remove visits.csv once DB is successfully updated
- [x] Switch to ruff for linting

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -24,22 +45,102 @@ def get_latest_release() -> models.LatestRelease:
...
+@asynccontextmanager
+async def lifespan(app: FastAPI):
+    # Initialise the DB and tables on server startup
     db.create_db_and_tables()
+    # Sync latest version tag using GitHub API
     app.latest_release = get_latest_release()
+    yield
+
+
+@repeat_every(seconds=15 * 60)  # every 15 minutes
```
Comment (ewels) [inline app/main.py:68]: ❤️

File: app/main.py [hunk 2]
```diff
@@ -24,22 +45,102 @@ def get_latest_release() -> models.LatestRelease:
...
+async def persist_visits():
+    """Write in-memory visits to a CSV file"""
+    global visit_data
+    with visit_data_lock:
+        if visit_data:
+            logger.debug(f"Persisting {len(visit_data)} visits to CSV {CSV_FILE_PATH}")
+            with open(CSV_FILE_PATH, mode="a") as file:
+                writer = csv.DictWriter(file, fieldnames=["timestamp"] + fieldnames)
+                for row in visit_data:
+                    writer.writerow(row)
```
Comment (ewels) [inline app/main.py]: Does the `.csv` file every get rotated / cleaned? Will it just grow and grow until it's GBs in size?

File: app/main.py [hunk 3]
```diff
@@ -24,22 +45,102 @@ def get_latest_release() -> models.LatestRelease:
...
+        df = pd.read_csv(CSV_FILE_PATH, sep=",", names=["timestamp"] + fieldnames)
+        df["start"] = pd.to_datetime(df["timestamp"])
+        df["end"] = df["start"] + pd.to_timedelta("1min")
+        df["start"] = df["start"].dt.strftime("%Y-%m-%d %H:%M")
+        df["end"] = df["end"].dt.strftime("%Y-%m-%d %H:%M")
+        df = df.drop(columns=["timestamp"])
+        # replace nan with "Unknown"
+        df = df.fillna("Unknown")  # df.groupby will fail if there are NaNs
+        # Summarize visits per user per minute
+        minute_summary = df.groupby(["start", "end"] + fieldnames).size().reset_index(name="count")
```
Comment (ewels) [inline app/main.py:145]: Does this do the entire CSV file every time? See above comment about cleaning that file, will the csv just grow and grow and this code get slower and slower?

**Review by ewels — COMMENTED** [review]
Couple of quick questions about how the CSV stuff works.. Looks good though 👍🏻

File: app/main.py (hunk 2 above)
Comment (vladsavelyev) [reply app/main.py]: It should be rotated on successful DB transaction - good point. Forgot to add a line to reset it 😅

File: app/main.py (hunk 3 above)
Comment (vladsavelyev) [reply app/main.py:145]: Yep, forgot to add the code to rotate it

File: app/db.py [hunk 4]
```diff
@@ -1,52 +1,50 @@
...
+assert sql_url is not None, sql_url
 engine = create_engine(sql_url)
 
 
-class Visit(SQLModel, table=True):  # type: ignore # mypy doesn't like this, not sure why
-    """Table to record raw individual visits to the version endpoint."""
+class Visits(SQLModel, table=True):  # type: ignore # mypy doesn't like this, not sure why
+    """
+    Table to record per-minute visit summaries. Start is a primary key,
+    and start and end are both indexed.
```
Comment (ewels) [inline app/db.py]:
```suggestion
    Table to record per-minute visit summaries.
    
    Start is a primary key, and start and end are both indexed.
```

Comment (ewels) [reply app/db.py]: Automatically generated docs will look nicer like this. Top line is the main description, everything after that is rendered as markdown docs under that. https://fastapi.tiangolo.com/tutorial/path-operation-configuration/?h=docstring#description-from-docstring

**Review by ewels — APPROVED** [review]

**Pull request merged**